src/crimson/os/seastore/lba_mapping.h

```diff
@@ -10,125 +10,198 @@
-  virtual laddr_t get_intermediate_base() const = 0;
-  virtual extent_len_t get_intermediate_length() const = 0;
-  // The start offset of the pin, must be 0 if the pin is not indirect
-  virtual extent_len_t get_intermediate_offset() const = 0;
+  laddr_t get_intermediate_key() const {
+    assert(!is_null());
+    assert(is_indirect());
+    return indirect_cursor->val.pladdr.get_laddr();
+  }
+  laddr_t get_intermediate_base() const {
+    assert(!is_null());
+    assert(is_indirect());
+    return physical_cursor->key;
+  }
+  extent_len_t get_intermediate_length() const {
+    assert(!is_null());
+    assert(is_indirect());
+    return physical_cursor->val.len;
+  }
+  // The start offset of the indirect cursor related to physical cursor
+  extent_len_t get_intermediate_offset() const {
+    assert(!is_null());
+    assert(is_indirect());
+    return get_intermediate_base().get_byte_distance<
+      extent_len_t>(get_intermediate_key());
+  }
 
-  virtual get_child_ret_t<lba_manager::btree::LBALeafNode, LogicalChildNode>
-  get_logical_extent(Transaction &t) = 0;
+  get_child_ret_t<lba_manager::btree::LBALeafNode, LogicalChildNode>
+  get_logical_extent(Transaction &t);
 
-  void link_child(LogicalChildNode *c) {
-    ceph_assert(child_pos);
-    child_pos->link_child(c);
+  void link_child(LogicalChildNode &extent) {
+    using lba_manager::btree::LBALeafNode;
+    auto leaf = physical_cursor->parent->cast<LBALeafNode>();
+    assert(leaf->is_stable());
+    leaf->link_child(&extent, physical_cursor->pos);
   }
-  virtual LBAMappingRef refresh_with_pending_parent() = 0;
 
-  // For reserved mappings, the return values are
-  // undefined although it won't crash
-  virtual bool is_stable() const = 0;
-  virtual bool is_data_stable() const = 0;
-  virtual bool is_clone() const = 0;
-  bool is_zero_reserved() const {
-    return !get_val().is_real();
+  LBAMapping duplicate() const {
+    assert(!is_null());
+    auto dup_iter = [](const LBACursorRef &iter) -> LBACursorRef {
+      if (iter) {
+	return std::make_unique<LBACursor>(*iter);
+      } else {
+	return nullptr;
+      }
+    };
+    return LBAMapping(dup_iter(physical_cursor), dup_iter(indirect_cursor));
   }
 
-  LBAMappingRef duplicate() const;
+private:
+  // Only allow BtreeLBAManager use these two methods
+  // FIXME: remove them
 
-  virtual ~LBAMapping() {}
-protected:
-  virtual LBAMappingRef _duplicate(op_context_t) const = 0;
+  friend class lba_manager::btree::BtreeLBAManager;
 
-  op_context_t ctx;
-  CachedExtentRef parent;
+  void make_indirect(LBACursorRef cursor) {
+    assert(physical_cursor);
+    assert(!indirect_cursor);
+    indirect_cursor = std::move(cursor);
+  }
 
-  pladdr_t value;
-  extent_len_t len = 0;
-  fixed_kv_node_meta_t<laddr_t> range;
-  uint16_t pos = std::numeric_limits<uint16_t>::max();
+  void link_physical(LBACursorRef cursor) {
+    assert(!physical_cursor);
+    assert(indirect_cursor);
+    physical_cursor = std::move(cursor);
+  }
```

Comment:
Let's add a, probably, `is_linked_direct()` method to help document interfaces that must be used after the linkage ?

Reply:
https://github.com/ceph/ceph/pull/62071#discussion_r1986761211 should be placed here.

`is_full_indirect()` should be equal to`is_linked_direct()`.

Reply:
I think `is_linked_direct()` here is equivalent to `is_full_indirect() || is_direct_mapping()` discussed above, which are the allowed states currently visible to the users.